Add lowering for resize, decomp for resize_as.

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121354
* __->__ #122317

This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @desertfire @chauhang

amjames committed:
Add lowering for resize, decomp for resize_as.

This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

[ghstack-poisoned]

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -2290,13 +2291,10 @@ def is_aligned(x):
 make_fallback(aten._cdist_backward)
-# See resize_storage_bytes
-make_fallback(aten.resize)
+# Looking like impossible, see: https://github.com/pytorch/pytorch/pull/121354
```
Move them to the very difficult/impossible part then.

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -5855,6 +5853,74 @@ def create_nn_parameter(self, placeholder):
+    assert isinstance(x, TensorBox)
+    assert isinstance(size, (list, tuple))
+
+    size = V.graph.sizevars.size_hints(size)
```
This is going to raise on unbacked symints. We should probably fallback in that case. Mind adding a test?

lezcano replied:
Actually, do you really need this line?

amjames replied:
You are right here we don't actually need this.

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -5855,6 +5853,74 @@ def create_nn_parameter(self, placeholder):
+        flat_index = out_indexer(idx)
+        flat_index_expr = ops.index_expr(flat_index, torch.int64)
+        limit = ops.index_expr(old_numel, torch.int64)
+        mask = ops.lt(flat_index_expr, limit)
+        x_val = flat_loader(
+            [
+                flat_index,
+            ]
+        )
+        return ops.where(mask, x_val, uninitalized_val)
```
Use `ops.masked`.

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -5855,6 +5853,74 @@ def create_nn_parameter(self, placeholder):
+        ranges=list(size),
+    )
+
+    return ir.ExternKernel.require_stride_order(
+        pointwise, ir.get_stride_order(new_stride)
+    )
```
Do we really need this?

amjames replied:
Without it the return always has contiguous stride order, which does not match the behavior of eager.

lezcano replied:
when exactly? Do you have a small repro or a test that fails?

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames commented on test/inductor/test_torchinductor.py:
```diff
@@ -8691,6 +8691,95 @@ def fn(input_ids) -> torch.Tensor:
+        self.assertTrue(same(correct_values, actual_values))
+        correct_strides = correct.stride()
+        actual_strides = actual.stride()
+        self.assertEqual(correct_strides, actual_strides)
```
@lezcano I never replied to [this comment](https://github.com/pytorch/pytorch/pull/122317#discussion_r1543430877) but this assert will fail if we don't work to make sure the output has the correct stride order. 

Specifically, cases where `memory_format` is either  `torch.channels_last` or  `torch.channels_last_3d` for both ops. In addition cases where `memory_format` is `torch.preserve_format` and `size_or_y.suggest_memory_format()` is `torch.channels_last` or `torch.channels_last_3d` for `resize_as` (in these cases `size_or_y` will be a Tensor).

lezcano commented:
Can we do this without realising the output, but rather changing the layout from Flexible to Fixed?

lezcano commented:
I've seen that whenever we use FixedLayout we always realise the input, so I guess that `as_strided` is effectively doing this.

@peterbell10 any thoughts?

FWIW, everything else looks reasonable to me.

peterbell10 commented:
> I've seen that whenever we use FixedLayout we always realise the input, so I guess that as_strided is effectively doing this.

A `Pointwise` object doesn't have a layout, only `Buffer` subclasses (and `ReinterpretView` of a buffer) have a layout. So realizing is the only way we can fix the strides.

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames commented:
@peterbell10 @lezcano I made one more update while you were looking at this.

It turns out we don't need to realize the pointwise *here*, but it is realized automatically in the code that handles the returned IR. In the same way the stride order does not need to be set after the fact. I use the `out_layout` indexer in the inner function this way it seems that the correct stride order is inferred.

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -5823,6 +5820,60 @@ def create_nn_parameter(self, placeholder):
+        # using zero as that is what empty does
+        uninitalized_val = 0.0
+
+    x_flat = view(x, (-1,))
```
Doesn't this require that `x` be contiguous? Otherwise the raw storage memory order won't match the flattened order.

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -5823,6 +5820,60 @@ def create_nn_parameter(self, placeholder):
+        flat_index_expr = ops.index_expr(flat_index, torch.int64)
+        limit = ops.index_expr(old_numel, torch.int64)
+        mask = ops.lt(flat_index_expr, limit)
+        return ops.masked(mask, lambda: flat_loader([flat_index]), uninitalized_val)
```
We should probably have a special case for `old_numel == 0` that just returns full of `uninitialized_val`.

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -5856,7 +5856,18 @@ def resize(x, size, *, memory_format=None):
         uninitalized_val = 0.0
 
-    x_flat = view(x, (-1,))
+    if V.graph.sizevars.evaluate_expr(sympy.Eq(old_numel, 0)):
```
This saves adding a guard:
```suggestion
    if V.graph.sizevars.statically_known_equals(old_numel, 0):
```

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -8691,6 +8691,116 @@ def fn(input_ids) -> torch.Tensor:
+    def test_resize(self):
+        def fn(x, size, memory_format):
+            # NOTE: Tensor.resize() =/= aten::resize()
+            return torch.ops.aten.resize(x, size, memory_format=memory_format)
```
btw do you know if functionalization generates this in normal codegen or do you have to explicitly call the aten funciton to hit this?

amjames replied:
`torch.resize` does not exist, and `Tensor.resize` is routed through [here](https://github.com/pytorch/pytorch/blob/aaef246c74b964ba43f051a4a3e484d25d418f44/torch/_tensor.py#L899-L905), which is implemented using a combination of `new`, `contiguous` and `view` [here](https://github.com/pytorch/pytorch/blob/aaef246c74b964ba43f051a4a3e484d25d418f44/torch/autograd/_functions/tensor.py#L31). There is a deprecation warning on those methods though. 

As far as I can tell you have to explicitly call the aten function to hit this.

peterbell10 replied:
Looks like functionalization does support functionalizing `aten.resize_` but it's not used in inductor.

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

lezcano reviewed (approved):
Nice, such a clean lowering!

lezcano reviewed (approved):
Nice, such a clean lowering!

amjames commented:
@pytorchbot merge

amjames committed:
Update on "Add lowering for resize, decomp for resize_as."


This has been split off from #121354 as the inplace version of these
methods prove to be rather tricky.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames commented:
@pytorchbot merge
